### PR TITLE
Update `findDocumentOrShadowRoot` to return the root node instead of throwing unnecessarily

### PR DESCRIPTION
**Pull request by ben10code**
**Description**
In Slate v0.62, a new util method `getDocumentOrShadowRoot` was introduced. This latter returns the `ShadowRoot` or `Document` node in which the editor is rendered but **throws** when not found. 

`Unable to find DocumentOrShadowRoot for editor element: [object HTMLDivElement]`

This seems to be causing issues for consumers who render the editor within React portals. The issue arises at unmount time, because `getRootNode` returns the now-detached portal node. 

Version `0.61.3` does not have this issue and seems to be the go-to for most right now.

**Issue**
Fixes: https://github.com/ianstormtaylor/slate/issues/4267

**Example**
In this first example, the editor is rendered inside a modal that uses React portal. When first opening the modal, it works fine - but closing the modal (unmounting) and reopening will crash because `root` isn't an instance of `Document` or `ShadowRoot` anymore.

https://codesandbox.io/s/slate-reproduction-reakit-portal-forked-yeykl?file=/index.js:902-907

In this second example, the portal is created using `createPortal` and then appended to the document later. In the initial render, the root node is not the `Document` and throws. (Example by JackRobards)

https://codesandbox.io/s/slate-reproduction-reakit-portal-fk002

**Context**
This pull request is straightforward and shouldn't introduce unwanted behaviours. Instead of throwing when the root isn't an instance of `Document` or `ShadowRoot`, I simply return `undefined`. In places where the root node is used, I simply check whether it exists before performing operations. 

I've been trying to gather context surrounding the reasons why we throw, but couldn't find anything. Feel free to share if you have context 🙏 ! 

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)
- [x] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

### CODE REVIEW

**Review by dylans — COMMENTED** [review]
Thank you @ben10code . This has very much been a problem for me as well in production and I was just about to tackle it so I really appreciate it. And the second commit really helped make this easier to review so thanks for that as well.

The general approach looks fine to me. I'm going to ask others on the Slate team for a quick review and hopefully we can land this quickly.

Also if you could add a changeset that would be great (helps us automate the release notes). You can run yarn changeset or in your branch and it will help you generate one. Thanks!

File: packages/slate-react/src/plugin/react-editor.ts [hunk 1]
```diff
@@ -111,9 +113,7 @@ export const ReactEditor = {
     if (el.ownerDocument !== document) return el.ownerDocument
 
     if (!(root instanceof Document || root instanceof ShadowRoot))
-      throw new Error(
-        `Unable to find DocumentOrShadowRoot for editor element: ${el}`
-      )
+      return undefined
```
Comment (dylans) [inline packages/slate-react/src/plugin/react-editor.ts:116]: Looking back at https://github.com/ianstormtaylor/slate/pull/3749/files , I think this may just suppress things, rather than restoring support for window/window.document. Perhaps better would be:

```suggestion
      return ReactEditor.getWindow(editor).document
```

And then we wouldn't need the change to support `| undefined` throughout? Unless I'm missing something (e.g. if window wasn't actually working previously).

File: packages/slate-react/src/components/android/android-editable.tsx [hunk 2]
```diff
@@ -96,7 +96,7 @@ export const AndroidEditable = (props: EditableProps): JSX.Element => {
       // Make sure the DOM selection state is in sync.
       const { selection } = editor
       const root = ReactEditor.findDocumentOrShadowRoot(editor)
-      const domSelection = root.getSelection()
+      const domSelection = root ? root.getSelection() : null
```
[the remark this answers is not in the document]
Comment (dylans) [reply packages/slate-react/src/components/android/android-editable.tsx:99]: Slate has started to use nullish coalescing and optional chaining so we should use them where appropriate. Assuming we don't go with my other suggestion.

File: packages/slate-react/src/plugin/react-editor.ts (hunk 1 above)
Comment (TheSpyder) [reply packages/slate-react/src/plugin/react-editor.ts:116]: This is the correct fix given what used to happen before #3749, and after it's merged every other change (to deal with potentially undefined return value) can be reverted.

I also suggest mentioning in the changeset that this fixes a regression caused by #3749.

File: packages/slate-react/src/plugin/react-editor.ts [hunk 3]
```diff
@@ -111,9 +111,7 @@ export const ReactEditor = {
     if (el.ownerDocument !== document) return el.ownerDocument
 
     if (!(root instanceof Document || root instanceof ShadowRoot))
-      throw new Error(
-        `Unable to find DocumentOrShadowRoot for editor element: ${el}`
-      )
+      return ReactEditor.getWindow(editor).document
```
Comment (clauderic) [inline packages/slate-react/src/plugin/react-editor.ts:114]: I think this be simplified to:

```suggestion
      return root.ownerDocument
```

Though to be honest, I think the entire method could be re-written, the logic is hard to follow currently.

Something like:

```ts
findDocumentOrShadowRoot(editor: ReactEditor): Document | ShadowRoot {
  const el = ReactEditor.toDOMNode(editor, editor)
  const root = el.getRootNode()

  if (root instanceof Document || root instanceof ShadowRoot) {
    return root
  }
  
  return root.ownerDocument
}
```

File: packages/slate-react/src/plugin/react-editor.ts (hunk 1 above)
Comment (ben10code) [reply packages/slate-react/src/plugin/react-editor.ts]: Good to go now!

File: packages/slate-react/src/plugin/react-editor.ts (hunk 3 above)
Comment (ben10code) [reply packages/slate-react/src/plugin/react-editor.ts]: Updated, indeed much cleaner! Also tested affected examples to ensure they still work :)

**Review by dylans — APPROVED** [review]

**Review by clauderic — APPROVED** [review]
